File: egottol/engines/eii/inference.py
import logging
from typing import Tuple

import numpy as np

from egottol.models.eii import DigitalHead, InferenceBackend, InferenceEngineConfig

logger = logging.getLogger(__name__)


def _softmax(logits: np.ndarray, temperature: float = 1.0) -> np.ndarray:
    t = max(temperature, 1e-9)
    shifted = logits / t - np.max(logits / t)
    exp = np.exp(shifted)
    return exp / np.sum(exp)


class InferenceEngine:
    """Inference operator Psi: embedding z -> prediction y_hat and confidence p."""

    def __init__(self, config: InferenceEngineConfig, weights: np.ndarray, bias: np.ndarray, conductance: np.ndarray):
        self.config = config
        self.weights = weights
        self.bias = bias
        self.conductance = conductance
        self._reservoir = None
        self._hopfield = None
        if self.config.backend == InferenceBackend.RESERVOIR:
            self._init_reservoir()
        elif self.config.backend == InferenceBackend.HOPFIELD:
            self._init_hopfield()

    def _init_reservoir(self) -> None:
        from egottol.engines.ai.reservoir import EchoStateReservoir, ReservoirConfig

        cfg = ReservoirConfig(
            n_reservoir=self.config.reservoir_size,
            spectral_radius=self.config.reservoir_spectral_radius,
            leak_rate=self.config.reservoir_leak,
            ridge_alpha=self.config.reservoir_ridge_alpha,
            input_scaling=self.config.reservoir_input_scaling,
            sparsity=self.config.reservoir_sparsity,
        )
        self._reservoir = EchoStateReservoir(cfg)
        if self.weights.size > 0 and self.weights.ndim == 2:
            dim = self.weights.shape[1]
            n = cfg.n_reservoir
            self._reservoir._ensure_input_weights(dim)
            win = np.zeros((n, dim))
            rows = min(n, dim)
            win[:rows, :rows] = np.eye(rows) * cfg.input_scaling
            self._reservoir.w_in = win
            readout = np.zeros((n, self.weights.shape[0]))
            readout[:rows, :] = self.weights[:, :rows].T
            self._reservoir._readout = readout

    def _init_hopfield(self) -> None:
        from egottol.engines.ai.hopfield_infer import HopfieldConfig, HopfieldInference

        patterns = self.config.hopfield_patterns
        if patterns is None and self.weights.size > 0:
            patterns = self.weights.tolist()
        if not patterns:
            dim = max(self.weights.shape[1] if self.weights.ndim == 2 else 4, 2)
            patterns = [np.ones(dim).tolist(), (-np.ones(dim)).tolist()]
        cfg = HopfieldConfig(
            max_iters=self.config.hopfield_max_iters,
            beta=self.config.hopfield_beta,
        )
        self._hopfield = HopfieldInference(np.asarray(patterns, dtype=float), cfg)

    def infer(self, z: np.ndarray) -> Tuple[np.ndarray, float]:
        z = np.asarray(z, dtype=float).reshape(-1)
        if self.config.backend == InferenceBackend.ANALOG:
            return self._infer_analog(z)
        if self.config.backend == InferenceBackend.ENERGY_BASED:
            return self._infer_energy_based(z)
        if self.config.backend == InferenceBackend.RESERVOIR:
            return self._infer_reservoir(z)
        if self.config.backend == InferenceBackend.HOPFIELD:
            return self._infer_hopfield(z)
        return self._infer_digital(z)

    def _infer_reservoir(self, z: np.ndarray) -> Tuple[np.ndarray, float]:
        if self._reservoir is None:
            self._init_reservoir()
        return self._reservoir.infer_embedding(z)

    def _infer_hopfield(self, z: np.ndarray) -> Tuple[np.ndarray, float]:
        if self._hopfield is None:
            self._init_hopfield()
        return self._hopfield.infer(z)

    def _infer_analog(self, z: np.ndarray) -> Tuple[np.ndarray, float]:
        g = self._match_conductance(z)
        v_row = self._dac_encode(z)
        i_col = g @ v_row
        y_hat = self._quantize_adc(i_col)
        if self.config.digital_head == DigitalHead.SOFTMAX:
            probs = _softmax(y_hat, self.config.temperature)
        else:
            probs = y_hat / max(np.sum(np.abs(y_hat)), 1e-9)
            probs = np.clip(probs, 0.0, None)
            probs = probs / max(np.sum(probs), 1e-9)
        confidence = float(np.max(probs))
        return probs, confidence

    def _infer_digital(self, z: np.ndarray) -> Tuple[np.ndarray, float]:
        w = self._match_weights(z)
        logits = w @ z + self.bias[: w.shape[0]]
        if self.config.digital_head == DigitalHead.SOFTMAX:
            probs = _softmax(logits, self.config.temperature)
        else:
            probs = _softmax(logits, self.config.temperature)
        confidence = float(np.max(probs))
        return probs, confidence

    def _infer_energy_based(self, z: np.ndarray) -> Tuple[np.ndarray, float]:
        w = self._match_weights(z)
        num_classes = w.shape[0]
        y = np.full(num_classes, 1.0 / num_classes, dtype=float)
        lr = self.config.ebm_learning_rate
        lam = self.config.ebm_lambda

        for _ in range(self.config.ebm_iterations):
            energy_grad = -w @ z + y
            y = y - lr * energy_grad
            y = np.clip(y, 0.0, None)
            y = y + lam * (1.0 - np.sum(y))
            total = np.sum(y)
            if total > 0:
                y = y / total

        confidence = float(np.max(y))
        return y, confidence

    def _dac_encode(self, z: np.ndarray) -> np.ndarray:
        z_norm = z / max(np.max(np.abs(z)), 1e-9)
        return self.config.v_dd * np.clip(z_norm, 0.0, 1.0)

    def _quantize_adc(self, currents: np.ndarray) -> np.ndarray:
        bits = self.config.adc_bits
        levels = 2**bits - 1
        scaled = currents / max(self.config.v_ref, 1e-9)
        quantized = np.round(np.clip(scaled, 0.0, 1.0) * levels) / levels
        return quantized

    def _match_weights(self, z: np.ndarray) -> np.ndarray:
        if self.weights.shape[1] == z.size:
            return self.weights
        w = np.zeros((self.weights.shape[0], z.size), dtype=float)
        cols = min(self.weights.shape[1], z.size)
        w[:, :cols] = self.weights[:, :cols]
        return w

    def _match_conductance(self, z: np.ndarray) -> np.ndarray:
        if self.conductance.shape[1] == z.size:
            return self.conductance
        g = np.zeros((self.conductance.shape[0], z.size), dtype=float)
        cols = min(self.conductance.shape[1], z.size)
        g[:, :cols] = self.conductance[:, :cols]
        return g

File: core/eii/inference.cpp
#include "inference.h"

#include <algorithm>
#include <cmath>
#include <numeric>

namespace deepiri {

std::vector<double> softmax(const std::vector<double>& logits, double temperature) {
    double t = std::max(temperature, 1e-9);
    std::vector<double> scaled(logits.size());
    for (size_t i = 0; i < logits.size(); ++i) scaled[i] = logits[i] / t;
    double maxVal = scaled.empty() ? 0.0 : *std::max_element(scaled.begin(), scaled.end());
    std::vector<double> exp(scaled.size());
    double sum = 0.0;
    for (size_t i = 0; i < scaled.size(); ++i) {
        exp[i] = std::exp(scaled[i] - maxVal);
        sum += exp[i];
    }
    if (sum <= 0.0) sum = 1e-9;
    for (double& v : exp) v /= sum;
    return exp;
}

std::pair<std::vector<double>, double> InferenceEngine::infer(const std::vector<double>& z) const {
    switch (config_.backend) {
        case InferenceBackend::Analog:
            return inferAnalog(z);
        case InferenceBackend::EnergyBased:
            return inferEnergyBased(z);
        case InferenceBackend::Digital:
        default:
            return inferDigital(z);
    }
}

Matrix InferenceEngine::matchWeights(size_t zSize) const {
    if (weights_.cols() == zSize) return weights_;
    Matrix w(weights_.rows(), zSize, 0.0);
    size_t cols = std::min(weights_.cols(), zSize);
    for (size_t i = 0; i < weights_.rows(); ++i) {
        for (size_t j = 0; j < cols; ++j) w.at(i, j) = weights_.at(i, j);
    }
    return w;
}

Matrix InferenceEngine::matchConductance(size_t zSize) const {
    if (conductance_.cols() == zSize) return conductance_;
    Matrix g(conductance_.rows(), zSize, 0.0);
    size_t cols = std::min(conductance_.cols(), zSize);
    for (size_t i = 0; i < conductance_.rows(); ++i) {
        for (size_t j = 0; j < cols; ++j) g.at(i, j) = conductance_.at(i, j);
    }
    return g;
}

std::vector<double> InferenceEngine::dacEncode(const std::vector<double>& z) const {
    double maxAbs = 1e-9;
    for (double v : z) maxAbs = std::max(maxAbs, std::abs(v));
    std::vector<double> out(z.size());
    for (size_t i = 0; i < z.size(); ++i) {
        double zn = z[i] / maxAbs;
        out[i] = config_.vDd * std::clamp(zn, 0.0, 1.0);
    }
    return out;
}

std::vector<double> InferenceEngine::quantizeAdc(const std::vector<double>& currents) const {
    double levels = static_cast<double>((1 << config_.adcBits) - 1);
    double vRef = std::max(config_.vRef, 1e-9);
    std::vector<double> out(currents.size());
    for (size_t i = 0; i < currents.size(); ++i) {
        double scaled = currents[i] / vRef;
        double clipped = std::clamp(scaled, 0.0, 1.0);
        out[i] = std::round(clipped * levels) / levels;
    }
    return out;
}

std::pair<std::vector<double>, double> InferenceEngine::inferAnalog(const std::vector<double>& z) const {
    Matrix g = matchConductance(z.size());
    std::vector<double> vRow = dacEncode(z);
    std::vector<double> iCol = g * vRow;
    std::vector<double> yHat = quantizeAdc(iCol);

    std::vector<double> probs;
    if (config_.digitalHead == DigitalHead::Softmax) {
        probs = softmax(yHat, config_.temperature);
    } else {
        double sumAbs = 1e-9;
        for (double v : yHat) sumAbs += std::abs(v);
        probs.resize(yHat.size());
        double sum = 0.0;
        for (size_t i = 0; i < yHat.size(); ++i) {
            double p = std::max(yHat[i] / sumAbs, 0.0);
            probs[i] = p;
            sum += p;
        }
        sum = std::max(sum, 1e-9);
        for (double& p : probs) p /= sum;
    }
    double confidence = probs.empty() ? 0.0 : *std::max_element(probs.begin(), probs.end());
    return {probs, confidence};
}

std::pair<std::vector<double>, double> InferenceEngine::inferDigital(const std::vector<double>& z) const {
    Matrix w = matchWeights(z.size());
    std::vector<double> logits = w * z;
    for (size_t i = 0; i < logits.size() && i < bias_.size(); ++i) logits[i] += bias_[i];
    // Both digital_head branches in inference.py call _softmax; ported verbatim.
    std::vector<double> probs = softmax(logits, config_.temperature);
    double confidence = probs.empty() ? 0.0 : *std::max_element(probs.begin(), probs.end());
    return {probs, confidence};
}

std::pair<std::vector<double>, double> InferenceEngine::inferEnergyBased(const std::vector<double>& z) const {
    Matrix w = matchWeights(z.size());
    size_t numClasses = w.rows();
    std::vector<double> y(numClasses, numClasses > 0 ? 1.0 / static_cast<double>(numClasses) : 0.0);
    std::vector<double> wz = w * z;

    for (int iter = 0; iter < config_.ebmIterations; ++iter) {
        std::vector<double> grad(numClasses);
        for (size_t i = 0; i < numClasses; ++i) grad[i] = -wz[i] + y[i];
        for (size_t i = 0; i < numClasses; ++i) y[i] -= config_.ebmLearningRate * grad[i];
        for (double& v : y) v = std::max(v, 0.0);
        double sumY = std::accumulate(y.begin(), y.end(), 0.0);
        for (double& v : y) v += config_.ebmLambda * (1.0 - sumY);
        double total = std::accumulate(y.begin(), y.end(), 0.0);
        if (total > 0.0) {
            for (double& v : y) v /= total;
        }
    }
    double confidence = y.empty() ? 0.0 : *std::max_element(y.begin(), y.end());
    return {y, confidence};
}

}  // namespace deepiri
